packages/worker/src/logger.ts

```diff
@@ -87,3 +87,68 @@ export class DefaultLogger implements Logger {
     this.log('ERROR', message, meta);
   }
 }
+
+export function withMetadata(logger: Logger, meta: LogMetadata | (() => LogMetadata)): Logger {
+  return new LoggerWithMetadata(logger, meta);
+}
+
+class LoggerWithMetadata implements Logger {
+  private parentLogger: Logger;
+  private metaChain: (LogMetadata | (() => LogMetadata))[];
+
+  constructor(parent: Logger, meta: LogMetadata | (() => LogMetadata)) {
+    // Flatten recusive LoggerWithMetadata instances
+    if (parent instanceof LoggerWithMetadata) {
+      this.parentLogger = parent.parentLogger;
+      this.metaChain = LoggerWithMetadata.appendToChain(parent.metaChain, meta);
```

Comment:
Can `this.parentLogger` be of type `LoggerWithMetadata`, i.e., more than one level deep... If not make a comment...

Reply:
I think the

```
    // Flatten recusive LoggerWithMetadata instances
    if (parent instanceof LoggerWithMetadata) {
```

lines just above already make this clear, no?

Reply:
I was referring to `this.parentLogger.parentLogger.`.... so the recursion can be several levels deep not just one